crates/typst-library/src/meta/figure.rs

```diff
@@ -393,6 +398,44 @@ impl FigureElem {
     }
 }
 
+/// The caption of a figure, including the supplement and counter.
+///
+/// This element is intended to be used in show rules to customize the
+/// appearance of the whole caption for all figure kinds.
+///
+/// ```example
+/// #show figure.caption: emph
+///
+/// #figure(
+///   rect(),
+///   caption: [A rectangle],
+/// )
+/// ```
+///
+/// When using a show rule for figures, you can wrap the full caption in a
+/// `{figure.caption}` element if you want it to be styled as other captions.
+///
+/// Display: Figure Caption
+/// Category: meta
+#[element(Show)]
+pub struct FigureCaption {
```

Comment:
I feel like the caption should have a few more fields which are populated by the figure's show rule, from which the user can assembly a caption themselves. Like `footnote.entry`.

Reply:
I more thought of still having it in the figure, but still providing it to the caption as a field of its own (just filled from the figure). But it's a bit tricky.

Reply:
This indeed seems tricky. If I understand well, you want a field that belongs to both a `figure` and its `figure.caption`?

Or maybe the field on caption would be `Smart`, with `auto` inheriting the value from its `figure`, and other values overriding it?

Reply:
I rather thought the figure would have it as `auto` and the caption doesn't have auto anymore because the figure already resolved it.

Reply:
Oh, you mean, in `FigureElem::show`, construct a `FigureCaption` with the resolved fields? Just want to make sure before implementing it.

Then, `FigureElem::full_caption` would have return a `figure.caption`, making `FigureElem::outline` return a `figure.caption` too. I don't know if this is a problem, but it is a side effect to consider: this means that show rules applying to `figure.caption`s would also apply in the outline.

Basically, if a `figure.caption` constructs its content itself instead of just acting as a proxy to a body that was constructed by its `figure`, we either:
1. Outline figures using with `figure.caption`s, making show rules apply to both the actual caption and in the outline, or
2. Outline figures by constructing separate, distinct, content.

Reply:
That's what I meant. I haven't thought deeply about it though and don't know whether it would work out well. The outline thing is definitely an important observation. Would be interested what @PgBiel and @Dherse think who've worked on the two things.

Reply:
I should also mention that I really like, in general, to have subelements for specific purpose such as a `figure.outline-entry` or `figure.outlined` as a shorthand for creating the outlines as it makes styling them very easy and is imo more composable than the existing technique

Reply:
> Oh, you mean, in `FigureElem::show`, construct a `FigureCaption` with the resolved fields? Just want to make sure before implementing it.

Yes, please refer to `outline.rs` for a sample implementation of the same idea (`OutlineElem`, containing general properties, constructs `OutlineEntry`, with specific per-item properties, for each entry).

> Then, `FigureElem::full_caption` would have return a `figure.caption`, making `FigureElem::outline` return a `figure.caption` too. I don't know if this is a problem, but it is a side effect to consider: this means that show rules applying to `figure.caption`s would also apply in the outline.

I think `full_caption` should only be used for the default, non-`show`-customized caption. The outline would, thus, use this, as _separate and distinct content_.

What I think here is that this `full_caption` method actually belongs to `FigureCaption`, perhaps to its `Show` implementation. Thus, we could just call some sort of `FigureCaption::new` from `FigureElem`, and then call `new_caption.show()` from `FigureElem::outline`, which will return its underlying content (thus avoiding problems with show rules).

>     2. Outline figures by constructing separate, distinct, content.
Therefore, I believe option 2 above is ideal.

Reply:
> I should also mention that I really like, in general, to have subelements for specific purpose such as a `figure.outline-entry` or `figure.outlined` as a shorthand for creating the outlines as it makes styling them very easy and is imo more composable than the existing technique

Not sure if I follow, but if you mean customizing how the figure is displayed in the outline, perhaps this could be redundant, as we already have outline.entry now, and could apply a show rule to entries based on whether `it.element.func() == figure` holds.

Reply:
This all sounds good. We should separate outline and caption and what @PgBiel says with the Show impl of FigureCaption also makes a lot of sense.

Reply:
I had forgotten we have `outline.entry`, see, I wasn't lying when I said I am tired. So, I agree with all of your points so far @PgBiel <3